[Sun Jia Cheng] iP

sunjc826 opened the pull request:
# Duke
## Features
*  CLI aka *keep* typing
*  **easy** to use

### Roadmap
1. Generic task creation
2. Task completion tracking
3. Deadlines and events

### Main method
```java
public class Duke {
  // more stuff...

  public static void main(String[] args) {
        run();
  }
}
```
### Progress
- [x] Todos, events, deadlines
- [ ] GUI  💯 
- [ ] ~~Commercial grade application~~

### Misc
- [Project repo](https://github.com/sunjc826/ip)
- `Here's some inline code`
Have an inspiration quote:
> Believe in your engineering capabilities as you would in a quote on the internet.
> - Sun Tzu

Commit:
Add Gradle support

sunjc826 committed:
Complete Level 1

sunjc826 committed:
Complete Level 2

Refactor modules to views.
Domain logic will be placed under domain directory.

sunjc826 committed:
Complete Level 3

Add Task class

sunjc826 committed:
Complete Level 4

Add subclasses of Task (Inheritance).
Add some minor error handling.

sunjc826 committed:
Improve bye message, and semi-automate testing

Other changes: Minor changes to runtest batch script

sunjc826 committed:
Complete Level 5

Add custom DukeException class

sunjc826 committed:
Complete Level 6

Remove some *lame* messages

sunjc826 committed:
Move all files into main java folder

sunjc826 committed:
Add documentation

sunjc826 committed:
Complete level 7

sunjc826 committed:
Complete level 8

sunjc826 committed:
Merge branch 'branch-Level-7'

sunjc826 committed:
Merge branch 'branch-Level-8'

sunjc826 committed:
fix localstorage bugs

sunjc826 committed:
Add root duke package

sunjc826 committed:
Add tests

sunjc826 committed:
Add JAR build configuration

sunjc826 committed:
Add documentation to public methods

sunjc826 committed:
Fix coding standard issues

sunjc826 committed:
Add find functionality

sunjc826 committed:
Merge branch 'branch-A-JavaDoc'

sunjc826 committed:
Merge branch 'branch-A-CodingStandard'

sunjc826 committed:
Merge branch 'branch-Level-9'

sunjc826 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

sunjc826 committed:
Add gradle, checkstyle and javafx

sunjc826 committed:
Add assertions

Code relies on if else conditionals to check for validity of states.

If else can make code look messier than it really is. Conditionals also
may not capture critical states that must hold true at certain places
in code.

Replacing conditionals with assertions at critical locations allows
us to force important conditions to hold. The program should crash
if these conditions do not hold, and simply printing to stdout is not
good enough.

Let's add assertions at critical locations like domain logic.

Assertions are preferable to conditionals for important logic when
checking for states that must hold.

sunjc826 committed:
Tasklist: add duplicate checks

Tasklist allows identical tasks (e.g. todos with the same name) to be
added.

This worsens the user experience as users need to keep track of
duplicate tasks by themselves.

Encapsulating the list of tasks in its own class allows
additional behavior to be added that is not provided by a regular
ArrayList. One of this is checking of duplicates when inserting into
the list.

Let's create a class called TaskList that encapsulates the
aforementioned behavior.

sunjc826 committed:
Add Level 10 documentation

sunjc826 committed:
Merge branch 'branch-Level-10'

feliciaivane commented on src/main/java/duke/domain/Task.java:
```diff
@@ -0,0 +1,120 @@
+     *
+     * @return A list of string fields representing the task.
+     */
+    public List<String> storageFields() {
```
Would it be better to name this method "getStorageFields" instead? 😄

sunjc826 replied:
Indeed, a verb would be better here!

feliciaivane commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,33 @@
+        chatbot.main();
+    }
+
+    public static void runGui(String[] args) {
```
I think there is a missing Javadocs header here?

sunjc826 replied:
Yup, thanks for the spot!

feliciaivane commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,33 @@
+import duke.views.cli.Cli;
+import duke.views.cli.strategies.MultiType;
+import duke.views.gui.Launcher;
+import javafx.application.Application;
```
Would it be better to move this import line above the other imports instead? 😬 Since you've put your java imports above your duke imports in your other classes. (Don't forget to add a blank line in between to keep the formatting consistent :D)

sunjc826 replied:
👍

sunjc826 replied:
Strangely, this is the way Checkstyle wants it. 🤔

feliciaivane commented on src/main/java/duke/views/cli/strategies/MultiType.java:
```diff
@@ -0,0 +1,238 @@
+        String typeString = fields[0];
+        Task task = null;
+        switch (typeString) {
+        case Todo.TYPE_STRING: {
```
I think the switch statement will still work fine without the brackets {} around each case 😄.

sunjc826 replied:
Yep, they can be removed.

feliciaivane commented on src/main/java/duke/views/cli/Cli.java:
```diff
@@ -0,0 +1,73 @@
+     */
+    @Override
+    public void listen() {
+
```
Would it be better to remove this empty line? :D I noticed this in some of your other methods too.

sunjc826 replied:
Yup, I'll do that.

feliciaivane reviewed (commented):
LGTM, just a few nits to fix. Good work :D

markuslim24 commented on src/main/java/duke/domain/Task.java:
```diff
@@ -0,0 +1,120 @@
+            this.setName(name);
+            typeString = (String) this.getClass().getField("TYPE_STRING").get(null);
+        } catch (NoSuchFieldException | IllegalAccessException e) {
+            System.out.println("Reflection error");
```
Would it be better to let another class handle the system printout of the error message here? Just to maintain the abstraction layer. Maybe could just use the PrintableMixin interface

sunjc826 replied:
Good idea! 
Edit: Yep, I'll make things consistent by using PrintableMixin more often.

markuslim24 commented on src/main/java/duke/views/cli/strategies/MultiType.java:
```diff
@@ -0,0 +1,238 @@
+
+    }
+
+    private String markDone(String query) throws DukeException {
```
Would breaking down the steps here and adding some comments help to improve readability here? Just a minor nitpick :P

sunjc826 replied:
Yep, in fact I definitely need to refactor these out into separate classes cos it's getting quite messy here.

markuslim24 commented on src/main/java/duke/shared/DateRange.java:
```diff
@@ -0,0 +1,148 @@
+     *
+     * @return Whether the date range is valid.
+     */
+    public boolean isValid() {
```
I wonder if having an isValid and isUnvalid function more confusing than just one of either? Just a personal thought haha

sunjc826 replied:
Yep I agree it is quite unnecessary. In fact it hasn't even been used. Just my ruby-ist mindset kicking in :)

markuslim24 reviewed (commented):
LGTM. Overall a clean codebase that is well-documented and quite readable. I've personally learnt a lot from this project as well on how I could better abstract some of my variables and methods. Awesome work man!

sunjc826 committed:
Make improvements based on review comments

sunjc826 committed:
Merge pull request #2 from sunjc826/branch-A-Assertions

Add assertions

sunjc826 committed:
Fix bugs

sunjc826 committed:
Merge pull request #1 from sunjc826/branch-C-DetectDuplicates

Tasklist: add duplicate checks

sunjc826 committed:
Refactor commands to separate class

MultiType class methods have become numerous and make the class messy.

Move each method to a separate Command subclass.

Extracting the method behavior to a separate command class reduces
coupling and makes the Responder class less bulky.

Let's create a Command class to encapsulate all commands that respond
to user input and generate a response string.

Encapsulation is better than storing all methods in a single class
as this improves code organization.

sunjc826 committed:
Merge pull request #3 from sunjc826/branch-A-CodeQuality

Refactor commands to separate class

sunjc826 committed:
Improve Gui

sunjc826 committed:
Merge branch 'branch-A-BetterGui'

sunjc826 committed:
Add documentation

sunjc826 committed:
Merge branch 'branch-A-UserGuide'

sunjc826 committed:
Fix README bugs

sunjc826 committed:
Cleanup code

sunjc826 committed:
Add more documentation

sunjc826 committed:
Disable assertions

sunjc826 committed:
Rename classes and improve documentation
